Fix mypy errors in openhands/llm directory

neubig opened the pull request:
This PR fixes all mypy errors in the openhands/llm directory by:

- Adding proper type annotations to functions
- Fixing ModelInfo type issues
- Fixing ModelResponse type issues
- Fixing async iterator type issues
- Adding proper return type annotations
- Fixing decorator type issues

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:ac07afb-nikolaik   --name openhands-app-ac07afb   docker.all-hands.dev/all-hands-ai/openhands:ac07afb
```

neubig commented on .github/workflows/py-unit-tests.yml:
```diff
@@ -48,7 +48,7 @@ jobs:
         run: make build
       - name: Run Tests
-        run: poetry run pytest --forked -n auto --cov=openhands --cov-report=xml -svv ./tests/unit --ignore=tests/unit/test_memory.py
+        run: poetry run pytest --forked --cov=openhands --cov-report=xml -svv ./tests/unit --ignore=tests/unit/test_memory.py
```
```suggestion
        run: poetry run pytest --forked -n auto --cov=openhands --cov-report=xml -svv ./tests/unit --ignore=tests/unit/test_memory.py
```

enyst commented on openhands/llm/retry_mixin.py:
```diff
@@ -12,25 +14,36 @@
+        num_retries = 3 if num_retries is None else num_retries
+        retry_min_wait = 1 if retry_min_wait is None else retry_min_wait
+        retry_max_wait = 60 if retry_max_wait is None else retry_max_wait
+        retry_multiplier = 2.0 if retry_multiplier is None else retry_multiplier
```
Sorry for the nit, this means that we set some default values manually again, when we already have the defaults set in [llm_config](https://github.com/All-Hands-AI/OpenHands/blob/70b21d16bd97ed98617ce0d97239123c4029040a/openhands/core/config/llm_config.py#L66)

It seems to me that sooner or later they'll be out of sync... On the other hand, the original code is not very readable either. Maybe a comment describing that those from config always fallback to the defaults in config could help?

neubig replied:
Fixed!

enyst commented on openhands/llm/llm.py:
```diff
@@ -305,17 +308,17 @@ def wrapper(*args, **kwargs):
             return resp
 
-        self._completion = wrapper
+        self._completion = partial(wrapper)
```
Just a question here, does this imply that during execution, it goes through another layer of indirection? I worry a bit that our code in this area is already not easy to follow, and while the majority of the changes in this PR make it more readable, I'm not really sure that wrapping it again does. 🤔

neubig replied:
I agree this was not good, fixed!

neubig commented:
@openhands-agent: Please take a look at @enyst 's comments and resolve them appropriately. I agree that we don't want to set defaults in two places, and it would be good if we could avoid the complexity posed by "partial".

openhands-agent commented:
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/13638117912).

neubig commented:
I have addressed the feedback:

1. Removed duplicate default values in `retry_mixin.py` since they are already defined in `LLMConfig`. Added a comment explaining where the default values come from.

2. Regarding the `partial()` usage - I initially tried to remove it to reduce complexity, but had to keep it because the type hints expect `partial` objects. The `partial()` usage here is actually necessary for proper type checking, as it ensures the wrapper functions maintain their correct types when assigned to class attributes.

neubig commented:
I have replaced `partial()` with `cast()` since we were not using `partial`'s argument binding functionality. This makes the code simpler while still satisfying the type checker.

For example, instead of:
```python
self._completion = partial(wrapper)
```

We now have:
```python
self._completion = cast('partial[ModelResponse]', wrapper)
```

This achieves the same type safety but without the extra indirection of `partial`.

neubig commented:
I have removed `partial()` since we were not using its argument binding functionality. Instead, I have:

1. Added proper type hints for class attributes:
```python
class LLM(RetryMixin, DebugMixin):
    _completion: Callable[..., ModelResponse]

class AsyncLLM(LLM):
    _async_completion: Callable[..., Coroutine[Any, Any, ModelResponse]]

class StreamingLLM(AsyncLLM):
    _async_streaming_completion: Callable[..., Coroutine[Any, Any, AsyncIterator[ModelResponse]]]
```

2. Added proper type hints for variables:
```python
async_completion_unwrapped: Callable[..., Coroutine[Any, Any, ModelResponse]] = self._async_completion
```

This makes the code clearer and more type-safe without the unnecessary indirection of `partial`.

enyst commented:
Integration tests for Deepseek failed on this branch, while they worked on `main` within the hour: https://github.com/All-Hands-AI/OpenHands/issues/5318#issuecomment-2695753558

It's Deepseek via its main provider, so it can always happen these weeks that it just fails. I can take a look later.

neubig commented:
@openhands

When using deepseek in integration tests, I get the following stack trace:
```
Traceback: Traceback (most recent call last):
  File "/home/runner/_work/OpenHands/OpenHands/openhands/controller/agent_controller.py", line 259, in _step_with_exception_handling
    await self._step()
  File "/home/runner/_work/OpenHands/OpenHands/openhands/controller/agent_controller.py", line 734, in _step
    action = self.agent.step(self.state)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/_work/OpenHands/OpenHands/openhands/agenthub/codeact_agent/codeact_agent.py", line 124, in step
    response = self.llm.completion(**params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pypoetry/virtualenvs/openhands-ai-2f2V0C4o-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 336, in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pypoetry/virtualenvs/openhands-ai-2f2V0C4o-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 475, in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pypoetry/virtualenvs/openhands-ai-2f2V0C4o-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 376, in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pypoetry/virtualenvs/openhands-ai-2f2V0C4o-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 398, in <lambda>
    self._add_action_func(lambda rs: rs.outcome.result())
                                     ^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/runner/.cache/pypoetry/virtualenvs/openhands-ai-2f2V0C4o-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 478, in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/home/runner/_work/OpenHands/OpenHands/openhands/llm/llm.py", line 268, in wrapper
    resp.choices[0], (ChatCompletion.Choice, ChatCompletionChunk.Choice)
                      ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pypoetry/virtualenvs/openhands-ai-2f2V0C4o-py3.12/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 264, in __getattr__
    raise AttributeError(item)
AttributeError: Choice
```

Please read the code, understand why this error might be happening, and fix it. You can download https://github.com/BerriAI/litellm and read the code to understand what the cause is.

neubig commented:
No we're getting a loop like this for deepseek...

```
2025-03-25 06:58:21,039 - INFO - Building image: ghcr.io/all-hands-ai/runtime:oh_v0.29.1_580hl6m5v4pwau75_ebmmjg6qh2jfa10z
2025-03-25 06:58:21,040 - INFO - [runtime 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Starting runtime with image: ghcr.io/all-hands-ai/runtime:oh_v0.29.1_580hl6m5v4pwau75_ebmmjg6qh2jfa10z
2025-03-25 06:58:21,748 - INFO - [runtime 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Container started: openhands-runtime-955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac. VSCode URL: None
2025-03-25 06:58:21,748 - INFO - [runtime 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Waiting for client to become ready at http://localhost:34231...
2025-03-25 06:58:37,792 - INFO - [runtime 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Runtime is ready.
2025-03-25 06:58:39,568 - INFO - [runtime 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Selected repo: None, loading microagents from /workspace/.openhands/microagents (inside runtime)
2025-03-25 06:58:39,571 - INFO - Loading user workspace microagents: []
2025-03-25 06:58:39,583 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.LOADING to AgentState.RUNNING
2025-03-25 06:58:39,615 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 0 GLOBAL STEP 0
2025-03-25 06:58:59,771 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 06:58:59,792 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 06:58:59,823 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 1 GLOBAL STEP 1
2025-03-25 06:59:47,171 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 06:59:47,192 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 06:59:47,223 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 2 GLOBAL STEP 2
2025-03-25 07:00:38,982 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:00:39,003 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:00:39,035 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 3 GLOBAL STEP 3
2025-03-25 07:01:30,214 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:01:30,235 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:01:30,266 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 4 GLOBAL STEP 4
2025-03-25 07:02:20,333 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:02:20,354 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:02:20,386 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 5 GLOBAL STEP 5
2025-03-25 07:03:14,403 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:03:14,424 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:03:14,456 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 6 GLOBAL STEP 6
2025-03-25 07:04:00,228 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:04:00,249 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:04:00,280 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 7 GLOBAL STEP 7
2025-03-25 07:04:54,083 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:04:54,105 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:04:54,137 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 8 GLOBAL STEP 8
2025-03-25 07:05:42,675 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:05:42,696 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:05:42,728 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 9 GLOBAL STEP 9
2025-03-25 07:06:38,282 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
2025-03-25 07:06:38,304 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING
2025-03-25 07:06:38,336 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] LEVEL 0 LOCAL STEP 10 GLOBAL STEP 10
2025-03-25 07:06:38,336 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.ERROR
2025-03-25 07:06:38,336 - ERROR - RuntimeError: Agent reached maximum iteration in headless mode. Current iteration: 10, max iteration: 10
2025-03-25 07:06:38,337 - WARNING - Stopping agent due to traffic control
2025-03-25 07:06:38,337 - INFO - [Agent Controller 955375b9-6b65-41d9-8a9e-ee60af1cb5be-57d01de3023bceac] Setting agent(CodeActAgent) state from AgentState.ERROR to AgentState.ERROR
2025-03-25 07:06:39,097 - INFO - Total events in history: 33
```

neubig commented:
Made a second simpler try in #7727
